### ISSUE TITLE
Introduce `ty_error`/`ty_error_with_message`/`const_error` to construct error type or const

### ISSUE THREAD

**Issue by mark-i-m** [rust-lang/compiler-team#297]
# What is this issue?

This is a **major change proposal**, which means a proposal to make a notable change to the compiler -- one that either alters the architecture of some component, affects a lot of people, or makes a small but noticeable public change (e.g., adding a compiler flag). You can read more about the MCP process on https://forge.rust-lang.org/.

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

# MCP Checklist

* [x] MCP **filed**. Automatically, as a result of filing this issue:
  * The @rust-lang/wg-prioritization group will add this to the triage meeting agenda so folks see it.
  * A Zulip topic in the stream `#t-compiler/major changes` will be created for this issue.
* [x] MCP **seconded**. The MCP is "seconded" when a compiler team member or contributor issues the `@rustbot second` command. This should only be done by someone knowledgable with the area -- before seconding, it may be a good idea to cc other stakeholders as well and get their opinion.
* [ ] **Final comment period** (FCP). Once the MCP is approved, the FCP begins and lasts for 10 days. This is a time for other members to review and raise concerns -- **concerns that should block acceptance should be noted as comments on the thread**, ideally with a link to Zulip for further discussion.
* [ ] MCP **Accepted**. At the end of the FCP, a compiler team lead will review the comments and discussion and decide whether to accept the MCP.
  * At this point, the `major-change-accepted` label is added and the issue is closed. You can link to it for future reference.
  
**A note on stability.** If your change is proposing a new stable feature, such as a `-C flag`, then a full team checkoff will be required before the feature can be landed. Often it is better to start with an unstable flag, like a `-Z` flag, and then move to stabilize as a secondary step.

# TL;DR

https://github.com/rust-lang/rust/pull/70551

- Make `TyKind::Error` and `ConstKind::Error` unconstructable outside `rustc_middle::ty`.
- Add `TyCtxt::ty_error*` and `TyCtxt::const_error` methods for constructing them instead. These methods automatically delay a span bug.

# Links and Details

*Add a few paragraphs explaining your design. The level of detail should be
sufficient for someone familiar with the compiler to understand what you're
proposing. Where possible, linking to relevant issues, old PRs, or external
documents like LLVM pages etc is very useful.*

PR: https://github.com/rust-lang/rust/pull/70551

`*::Error` variants are intended for usage _only_ when we know the compilation session will fail. In this case, we can use them to prevent emitting further errors and overwhelming the user. If we use it and compilation does _not_ fail, we may accidentally allow bad code. So far, this has been enforced by convention "just do the right thing".

This PR enforces the invariant programmatically: it adds a private field to `TyKind::Error` and `ConstKind::Error` so that they are unconstructable outside of `rustc_middle::ty`. It then adds the aforementioned methods on `TyCtxt` to construct the errors. These methods automatically delay a span bug so that the use of an `*::Error` now guarantees compilation failure.

A lot of locations already call `delay_span_bug` before returning a `*::Error`. To avoid redundant delayed span bugs, there is a `ty_error_with_message` that takes a span and message. In most cases, the existing `delay_span_bug` can just use that instead.

As an added bonus, that PR also adds `track_caller` to `delay_span_bug` and the `ty_error*` methods. In this case, it means that we get the location where the spurious `*::Error` was constructed.

# Mentors or Reviewers

*Who will review this work? If you are being mentored by someone, then list
their name here. If you are a compiler team member/contributor, and you
intend to mentor someone else, then you can put your own name here. You can
also leave it blank if you are looking for a reviewer. (Multiple names are ok
too.)*

mentor/reviewer: @eddyb

**Comment by nikomatsakis** [comment]
For the record I think this idea is great!

**Comment by mark-i-m** [comment]
@nikomatsakis  So it's been about 10 days... is there something else that needs to be done here?

**Comment by nikomatsakis** [comment]
Nope. Once the FCP expires and no complaints were raised, the MCP is accepted.
